In [2]:
import numpy as np
import matplotlib.pyplot as plt
import scipy.stats as st

In [3]:
y1 = np.array([70, 50, 65, 60, 75, 67, 74])
y2 = np.array([80, 74, 90, 70, 75, 65, 85])
y3 = np.array([148, 142, 140, 150, 160, 170, 155])

n = 21
k = 3


для определения межгруповой дисперсии нужны средние значения по группам

In [4]:
y_mean1 = np.mean(y1)
y_mean2 = np.mean(y2)
y_mean3 = np.mean(y3)

y_mean1, y_mean2, y_mean3

(65.85714285714286, 77.0, 152.14285714285714)

In [5]:
total = np.array([y1, y2, y3])
total

array([[ 70,  50,  65,  60,  75,  67,  74],
       [ 80,  74,  90,  70,  75,  65,  85],
       [148, 142, 140, 150, 160, 170, 155]])

In [6]:
y_mean_total = np.mean(total)
y_mean_total

98.33333333333333

Сумма квадратов отклонений наблюдений от общего среднего

In [7]:
np.sum((total - y_mean_total)**2) # S общ^2

32400.66666666667

Сумма квадратов отклонений средних групповых значений от общего среднего

In [8]:
S_f = np.sum(((y_mean1 -y_mean_total)**2) + ((y_mean2 -y_mean_total)**2) + ((y_mean3 -y_mean_total)**2)) * 7 #число наблюдений в каждой группе равно семи
S_f # S - факторное

30836.952380952374

Остаточная сумма квадратов отклонений

In [9]:
S_ost = np.sum(((y1 -y_mean1)**2) + ((y2 -y_mean2)**2) + ((y3 -y_mean3)**2))
S_ost

1563.7142857142856

In [10]:
S_f + S_ost

32400.66666666666

In [11]:
D_f = S_f / (k - 1)
D_f # факторная дисперсия

15418.476190476187

In [12]:
D_ost = S_ost / (n - k)
D_ost #остаточная дисперсия

86.87301587301586

Наблюдаемо значение критерия ФИшера

In [13]:
F_n = D_f / D_ost
F_n

177.48291613374747

Сравнить наблюдаемое значение с табличным

F_t = 3.55смо трим в таблице относительно степенй свободы числителя и знаменателя

F_n > F_t -> H1 статистически значимые различия есть


In [14]:
f = st.f_oneway(y1, y2, y3)
f #функция однофакторного дисперсионного анализа

F_onewayResult(statistic=177.48291613374704, pvalue=1.420466900107174e-12)

statistic=177.48291613374704 -> F_n
pvalue=1.420466900107174e-12 очень маленькое < alpha -> H1

Post hoc tes Tukey

In [15]:
from statsmodels.stats.multicomp import pairwise_tukeyhsd
import pandas as pd

In [16]:
df = pd.DataFrame({'score': [70, 50, 65, 60, 75, 67, 74,
                            80, 74, 90, 70, 75, 65, 85,
                            148, 142, 140, 150, 160, 170, 155],
                'group' : np.repeat(['accounter', 'lawyer', 'programmer'], repeats=7)})
tukey = pairwise_tukeyhsd(endog =df['score'],
                        groups= df['group'],
                        alpha = 0.05)
print(df)
print(tukey)

    score       group
0      70   accounter
1      50   accounter
2      65   accounter
3      60   accounter
4      75   accounter
5      67   accounter
6      74   accounter
7      80      lawyer
8      74      lawyer
9      90      lawyer
10     70      lawyer
11     75      lawyer
12     65      lawyer
13     85      lawyer
14    148  programmer
15    142  programmer
16    140  programmer
17    150  programmer
18    160  programmer
19    170  programmer
20    155  programmer
    Multiple Comparison of Means - Tukey HSD, FWER=0.05    
  group1    group2   meandiff p-adj   lower   upper  reject
-----------------------------------------------------------
accounter     lawyer  11.1429 0.0918 -1.5722 23.8579  False
accounter programmer  86.2857    0.0 73.5707 99.0007   True
   lawyer programmer  75.1429    0.0 62.4278 87.8579   True
-----------------------------------------------------------


accounter     lawyer  11.1429 0.0918 -1.5722 23.8579  False различия не найдены 
pvalue (0.0918) > alpha (0.05)

accounter programmer  86.2857    0.0 73.5707 99.0007   True
   lawyer programmer  75.1429    0.0 62.4278 87.8579   True

различия найдены верна альтернативная гипотеза

Двухфакторный дисперсионный анализ

In [17]:
y111 = 57
y112 = 59
y11 = (y111 + y112)/2
y11

58.0

In [18]:
y121 = 56
y122 = 58
y12 = (y121 + y122)/2
y12

57.0

In [19]:
y211 = 32
y212 = 34
y21 = (y211 + y212)/2
y21

33.0

In [20]:
y221 = 71
y222 = 71
y22 = (y221 + y222)/2
y22

71.0

In [21]:
y_mean_A1 = (y11 + y12)/2
y_mean_A2 = (y21 + y22)/2
y_mean_B1 = (y11 + y21)/2
y_mean_B2 = (y12 + y22)/2

y_mean_A1, y_mean_A2, y_mean_B1, y_mean_B2

(57.5, 52.0, 45.5, 64.0)

In [22]:
y_mean = np.mean(y_mean_A1 + y_mean_A2 + y_mean_B1 + y_mean_B2)/4
y_mean

54.75

## ANOVA

SSt = SSA + SSB + SSAB +SSE сумма кватратов отклонений общая = сумма квадратов отклонений воздействия фактора А + сумма квадратов откланений фактора В + сумма квадратов отклонений взамидейсвия факторов АВ + сумма квадратов отклонеия случайной ошибки
SSt  = sum((yijk)**2) - a*b*n*(y_mean**2) нужно  из суммы квадратов каждого значения вычесть произведение а - количество уровней фактоа А(2), b -  количество уровней фактоа B(2), n - количество репликаций(опытов, повторений)
SSt = 1511.5

сумма квадратов отклонений воздействия фактора А
SSA = a*n*((y_mean_A1**2)+(y_mean_A2**2)) - a*b*n*(y_mean)**2 или a*n*sum((y_mean_A)**2) - a*b*n*(y_mean)**2 = 60.5

сумма квадратов отклонений воздействия фактора B
SSB= a*n*((y_mean_B1**2)+(y_mean_B2**2)) - a*b*n*(y_mean)**2 или 
a*n*sum((y_mean_B)**2) - a*b*n*(y_mean)**2 = 684.5

квадратов отклонений взамидейсвия факторов АВ
SSAB = n * (sum((yij_mean)**2)) - a*b*n*(y_mean**2) - SSA - SSB = 
n * ((y_meanA1)**2) + (y_meanA2)**2) + (y_meanB1)**2) + (y_meanB2)**2)) - a*b*n*(y_mean**2) - SSA - SSB = 760.5

сумма квадратов отклонеия случайной ошибки
SSE = SSt - SSA - SSB - SSAB = 1511.5 - 60.5 - 684.5 - 760.5 = 6



In [23]:
a = 2 #уровень фактора А
b = 2 #уровень фактора В
n = k = 2 # число репликаций

dfA = (a-1) #степени свободы фактора А
dfB = (b-1) 
dfAB = (a-1) * (b-1) 
dfE = a * b * (n-1)


Расчитать сколько квадратных отклонений приходится на одну степень свободы

In [24]:
SSA = 60.5
SSB = 684.5
SSAB = 760.5
SSE = 6

MSA = SSA / dfA
MSB = SSB / dfB
MSAB = SSAB / dfAB
MSE = SSE / dfE

MSA, MSB, MSAB, MSE

(60.5, 684.5, 760.5, 1.5)

Считаем критерий Фишера

In [25]:
FA = MSA / MSE
FB = MSB / MSE
FAB = MSAB / MSE

F_t = 7.71 #табличное

FA, FB, FAB

(40.333333333333336, 456.3333333333333, 507.0)

все значения  > табличного(критического) -> H1, статичстически значимые различия найдены

In [26]:
import statsmodels.api as sm
from statsmodels.formula.api import ols



In [27]:
fA = np.array(["low","low","low","low","high","high","high","high"])
fA

array(['low', 'low', 'low', 'low', 'high', 'high', 'high', 'high'],
      dtype='<U4')

In [28]:
fB = np.array(["low","low","high","high","low","low","high","high"])
fB

array(['low', 'low', 'high', 'high', 'low', 'low', 'high', 'high'],
      dtype='<U4')

In [29]:
values = np.array([57, 59, 56, 58, 32, 34, 71, 71])
values

array([57, 59, 56, 58, 32, 34, 71, 71])

In [30]:
dp = pd.DataFrame({'fA': fA, 'fB':fB, 'values':values})
dp

fA    fB  values
0   low   low      57
1   low   low      59
2   low  high      56
3   low  high      58
4  high   low      32
5  high   low      34
6  high  high      71
7  high  high      71

In [31]:
строим ANOVA таблицу

SyntaxError: invalid syntax (2535720959.py, line 1)

In [ ]:
#строим модель при помощи метола ols

lm_model = ols('values ~ C(fA) * C(fB)', data=dp).fit()

#строим ANOVA

table = sm.stats.anova_lm(lm_model, type = 2)
table

df  sum_sq  mean_sq           F    PR(>F)
C(fA)        1.0    60.5     60.5   40.333333  0.003150
C(fB)        1.0   684.5    684.5  456.333333  0.000028
C(fA):C(fB)  1.0   760.5    760.5  507.000000  0.000023
Residual     4.0     6.0      1.5         NaN       NaN

              df  sum_sq  mean_sq           F    PR(>F)
C(fA)        1.0    60.5     60.5   40.333333  0.003150
C(fB)        1.0   684.5    684.5  456.333333  0.000028
C(fA):C(fB)  1.0   760.5    760.5  507.000000  0.000023
Residual     4.0     6.0      1.5         NaN       NaN

получаем все наши значения

PR(>F) - это pvalue
507.000000  0.000023 Самый большой показатель критерия фишера получился для взаимодействия факторов и тут самы маленький pvalue -> следовательно самое большое влияние оказывает взаимодействие факторов.